### ISSUE TITLE
dependencies of file: or link: local deps are not installed

### ISSUE THREAD

**Issue by wclr** [pnpm/pnpm#3034]
### pnpm version: 5.13.6

Well not sure why this workflow is not supported. Can be configured as workspaces, but wanted to try without.

I have a folder inside a project which I add to root package.json as `link:` (or `file:`) dependency. This package inside its `package.json` has `dependencies`. But they are not installed when running `pnpm install` in the root.

I tried to add this local folder dep with a `pnpm add` (not manually) but the result is the same.

### Expected behavior:

Deps of `file:`/`link:` dependency package are installed.

### Actual behavior:

Deps of `file:`/`link:` dependency package are not installed.

**Comment by wclr** [comment]
@zkochan can you comment on that?

**Comment by zkochan** [comment]
Could you just make a workspace from it? Then it will work as you'd expect.

**Comment by wclr** [comment]
So if I want to have any `file:` dependency (so that pnpm would effect it's inner node_modules) a project it has to be a workspace that includes this dependency? Correct?

**Comment by zkochan** [comment]
Yes, correct at the moment.

Alex <notifications@github.com> ezt írta (időpont: 2021. jan. 9., Szo
14:22):

**Comment by wclr** [comment]
Should it be closed or this is a subject for change?

**Comment by zkochan** [comment]
I don't know, it might be useful to at least print an info message in this case.

**Comment by wclr** [comment]
Could you still elaborate on the logic behind this behavior? Does pnpm in case of `file:` dependency expects that this package is something external and already should have installed dependencies inside?

**Comment by zkochan** [comment]
Correct, this is how it currently works. It will not install dependencies in external projects.

**Comment by wclr** [comment]
NPM seems to treat it a little bit differently with `file:` deps, it makes symlink in node_modules and installs its dependencies, if a pack has a dependency that can not be hoisted it installs it inside of `file:` package location.

**Comment by farfromrefug** [comment]
@zkochan is there any workaround for this? It is a pretty big issue. If you have a repo setup with `file:` because it works with npm and yarn, then you simply cant use `pnpm` because of this :s 
We are now using that setup for around 30 github repo because we need to. But because of this i cant use pnpm anymore :s

**Comment by zkochan** [comment]
You can try to use it with the [dependenciesMeta.injected](https://pnpm.io/package_json#dependenciesmetainjected). It will "hard link" the external package as a regular dependency and install its dependencies.

**Comment by farfromrefug** [comment]
@zkochan thanks i tried and it kind of works. Strangely  not all deep deps are hard linked. 
this is the repo https://github.com/nativescript-community/ui-pager
the root package.json as the `tools` folder as a dep.
if i had 
```
  "dependenciesMeta": {
        "@nativescript-community/plugin-seed-tools": {
          "injected": true
        }
      }
```
then some deep deps get install liked typescript. But weirldy others like `ts-patch` does not get hard linked in node_modules. Nor does its binary linked in `.bin` though it is in the `tools` `package.json`

EDIT: it works if i use `shamefully-hoist` but i should not have to in this case right?

**Comment by farfromrefug** [comment]
@zkochan do you have any idea on this? if i have a dependency like this `"@nativescript-community/template-snippet": "file:../demo-snippets"` then running `pnpm i` wont initialize the node_modules folder in `../demo-snippets`

**Comment by farfromrefug** [comment]
@zkochan could we have a `.npmrc` setting to force `file:` to be treated as normal dependencies? I have `file:` which requires other `file:` deps and using [dependenciesMeta.injected](https://pnpm.io/package_json#dependenciesmetainjected) gets too messy

**Comment by zkochan** [comment]
@farfromrefug it works the way you want it to work in pnpm v7

**Comment by farfromrefug** [comment]
@zkochan awesome ! i ll try v7 and report

**Comment by farfromrefug** [comment]
@zkochan i can confirm everything works perfectly in v7. Thanks for your amazing work!

**Comment by JulesAaelio** [comment]
Hello @zkochan, 
I'm using pnpm v7.9.3. 

I have a `:file` dependency with `:file` that has multiple `file:` dependencies as well. 
When I run `pnpm install` the package is symlinked in the node_modules and its dependencies are not installed. 

I think it's the same use case as @farfromrefug 
> @zkochan could we have a `.npmrc` setting to force `file:` to be treated as normal dependencies? I have `file:` which requires other `file:` deps and using [dependenciesMeta.injected](https://pnpm.io/package_json#dependenciesmetainjected) gets too messy

Which is supposed to work with pnpm v7
> @farfromrefug it works the way you want it to work in pnpm v7

But it does not work for me, am I missing something ? 
Note that I also tried with  [dependenciesMeta.injected](https://pnpm.io/package_json#dependenciesmetainjected) and obtained the same result.

**Comment by blowsie** [comment]
I also have this issue when using pnpm link 
pnpm v7.6.0

**Comment by bgrand-ch** [comment]
My solution with Firebase pre-deploy and post-deploy hooks, maybe it can help you:

https://stackoverflow.com/questions/75785161/how-to-use-pnpm-workspace-protocol-and-deploy-firebase-functions

**Comment by Aeolun** [comment]
`file:` dependencies do nothing for me in `v9.9.0`. What I think is infinitely more annoying is that pnpm doesn't complain about it. It just does nothing. Going from one version to a local link leaves the existing version in the node_modules and acts as if install has succeeded.

**Comment by jdalton** [comment]
I'll throwing my hat into the "doesn't work" ring. I've noticed errors when shipping packages with overrides that use "file:".
>  ERR_PNPM_LINKED_PKG_DIR_NOT_FOUND  Could not install from "/path/to/socketregistry/packages/manual/assert/overrides/call-bind" as it does not exist.

The workaround is publishing a stub to npm with a specific release version. It "just works" with npm/yarn.

**Comment by EfeDurmaz16** [comment]
@zkochan, I checked the protocol cases separately using rebuilt v11.27.0 and v12.5.1 CLIs from main commit `1be9010902aeaa5aa4b7376fd745e8f848eb48c9` on macOS arm64.

- `file:./lib` installed the local package's dependencies. Nested `file:` dependencies and switching an existing registry dependency to `file:` also passed in the tested fixtures.
- `link:` and `pnpm link` did not install dependencies inside the linked source project. That matches the [documented link contract](https://pnpm.io/cli/link#whats-the-difference-between-pnpm-link-and-using-the-file-protocol).
- Bare `pnpm add ./lib` behaved like linking, as documented under [local package sources](https://pnpm.io/package-sources#local-file-system).

I could not reproduce the `file:` failure in those cases. The later overrides and package-specific reports may have other conditions, so I would not treat this as proof that the whole thread is resolved. A minimal current `file:` reproducer would help separate a remaining bug from the expected `link:` behavior.

Written by an agent (Codex, GPT-6).

**Comment by zkochan** [comment]
Right, this must be fixed and link: works as per documentation.
